apps/build-onchain-apps/contracts/src/SignatureMintERC721.sol

```diff
@@ -0,0 +1,143 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.13;
+
+import {ERC721} from "@openzeppelin/contracts/token/ERC721/ERC721.sol";
+import {Strings} from "@openzeppelin/contracts/utils/Strings.sol";
+import "@openzeppelin/contracts/utils/cryptography/ECDSA.sol";
+import "@openzeppelin/contracts/utils/Address.sol";
+import "@openzeppelin/contracts/utils/cryptography/MessageHashUtils.sol";
+import {NFTUtilities} from "./NFTUtilities.sol";
+
+// Errors
+error InsufficientFunds();
+
+/**
+ * @title SignatureMintERC721
+ * @dev contract that allows a user to mint a ERC721 for free with a cryptographically signed message.
+ *      This is useful for mints where you want to allow users to mint for free based on a signature from
+ *      your backend API.  This is an alternative approach to a merkel tree which is fully on-chain.
+ */
+contract SignatureMintERC721 is ERC721 {
+    using ECDSA for bytes32;
+    using MessageHashUtils for bytes32;
+
+    // Contract Params
+    uint256 public paidMintCost = 0.0001 ether;
+    address public freeMintSigner; // Wallet Address for the authorized wallet to distribute free mints.
+    uint256 public currentId = 0; // Current NFT id
+    mapping(address => uint256) public usedFreeMints; // Mapping to store free mints
+
+    // Setup contract Metadata
+    string public imageUri = "ipfs://QmQRzw5MtNQcdjg9rZ8MYugAz6WjaBdvsuNShKTLYNJdj7/coffee_dog.png";
+    NFTUtilities.CollectionMetadata public metadata =
+        NFTUtilities.CollectionMetadata("CoffeeDog", "Buy a dog a coffee", "CDOG", imageUri, "");
+
+    // Events
+    event Mint(address indexed buyer, uint256 tokenId, uint256 cost);
+
+    constructor(address signer) ERC721(metadata.name, metadata.symbol) {
+        freeMintSigner = signer;
+    }
+
+    /**
+     * @dev Function to mint for free using a wallet signature
+     * @param  account that will receive the NFT
+     * @param  signature that authorizes this transaction to mint for free.
+     */
+    function freeMint(address account, bytes memory signature) external {
+        require(this.verifySignature(account, signature) == true, "Unable to verify Signature");
+
+        // Check if the free mint has already been used (1 per wallet)
+        require(usedFreeMints[account] == 0, "Free mint already used");
+        usedFreeMints[account] = 1;
+
+        // Mint the NFT
+        _mint(account, ++currentId);
+        emit Mint(account, currentId, 0);
```

Comment:
Normally transfers from the 0 address are indexed as mints and there is no special mint event. Do you need it?

Reply:
I thought it would be nice to showcase how a custom event would work.. Also, I wanted to be able to show the mint event + price easily on the UI.  I know you can get that with transaction logs, just requires a bit more work.